toolchains/prebuilt_toolchains.py

```diff
@@ -0,0 +1,370 @@
+                    sha256=sha256,
+                    prefix=prefix,
+                    url=CMAKE_URL_TEMPLATE.format(
+                        full=version,
+                        file=file
+                    ),
+                    build="cmake",
+                    template="_CMAKE_BUILD_FILE",
+                    bin=bin,
+                    exec_compatible_with=CMAKE_TARGETS[plat_target],
+                )
+            )
+            version_toolchains.append("@{}//:cmake_toolchain".format(name))
+
+        archives.append("\n".join(
+            [
+                "    if \"{}\" == version:".format(version),
+            ] + [indent(archive, " " * 8) for archive in version_archives])
+        )
+
+        archives.append(indent(REGISTER_TOOLCHAINS.format(
+            toolchains="\n".join(
+                [indent("\"" + toolchain + "\",", " " * 4) for toolchain in version_toolchains])
+        ), " " * 8))
+
+        archives.extend([
+            indent("return", " " * 8),
+            "",
+        ])
+
+    archives.append(
+        indent("fail(\"Unsupported version: \" + str(version))", " " * 4))
+
+    return "\n".join([archive.rstrip(" ") for archive in archives])
+
+
+def get_ninja_definitions() -> str:
+    """Define a set of repositories and calls for registering `ninja` toolchains
+
+    Returns:
+        str: The Implementation of `_ninja_toolchains`
+    """
+
+    archives = []
+
+    for version in NINJA_VERSIONS:
+
+        version_archives = []
+        version_toolchains = []
+
+        for target in NINJA_TARGETS.keys():
+            url = NINJA_URL_TEMPLATE.format(
+                full=version,
+                target=target,
+            )
+
+            # Get sha256 (can be slow)
+            remote = urllib.request.urlopen(url)
+            total_read = 0
+            max_file_size = 100*1024*1024
+            hash = hashlib.sha256()
+            while True:
+                data = remote.read(4096)
+                total_read += 4096
+
+                if not data or total_read > max_file_size:
+                    break
+
+                hash.update(data)
+            sha256 = hash.hexdigest()
+
+            name = "ninja_{}_{}".format(version, target)
+
+            version_archives.append(
+                REPO_DEFINITION.format(
+                    name=name,
+                    url=url,
+                    sha256=sha256,
+                    prefix="",
+                    build="ninja",
+                    template="_NINJA_BUILD_FILE",
+                    bin="ninja.exe" if "win" in target else "ninja",
+                    exec_compatible_with=NINJA_TARGETS[target],
+                )
+            )
+            version_toolchains.append("@{}//:ninja_toolchain".format(name))
+
+        archives.append("\n".join(
+            [
+                "    if \"{}\" == version:".format(version),
+            ] + [indent(archive, " " * 8) for archive in version_archives])
```

Comment:
Ditto as above

Reply:
Moving conversation to https://github.com/bazelbuild/rules_foreign_cc/pull/508#discussion_r576668650